engine/src/board.rs

```diff
@@ -127,169 +87,138 @@ impl Board {
                 reason: "Trying to move a covered piece".to_string(),
             });
         }
-        if let Some(vec) = self.board.get_mut(current) {
-            if let Some(piece) = vec.pop() {
-                if vec.is_empty() {
-                    self.board.remove(current);
-                }
-                self.insert(target, piece);
-                return Ok(());
-            }
-        }
-        panic!(
-            "Trying to move {piece} from {current} to {target} which should have been a legal move"
-        );
+        let bug_stack = self.board.get_mut(current);
+        let piece = bug_stack.pop_piece();
+        self.insert(target, piece);
+        Ok(())
     }
 
-    pub fn neighbor_is_a(&self, position: &Position, bug: Bug) -> bool {
+    pub fn neighbor_is_a(&self, position: Position, bug: Bug) -> bool {
         self.top_layer_neighbors(position)
             .iter()
-            .any(|piece| piece.bug == bug)
+            .any(|piece| piece.bug() == bug)
     }
 
-    pub fn position(&self, piece: &Piece) -> Option<Position> {
-        self.board.iter().find_map(|(pos, pieces)| {
-            if pieces.contains(piece) {
-                Some(*pos)
-            } else {
-                None
-            }
-        })
+    pub fn level(&self, position: Position) -> usize {
+        self.board.get(position).size as usize
     }
 
-    pub fn piece_already_played(&self, piece: &Piece) -> bool {
-        self.board.values().any(|pieces| pieces.contains(piece))
-    }
-
-    pub fn positions_for_color(&self, color: &Color) -> Vec<Position> {
-        self.board
-            .iter()
-            .filter_map(|(pos, pieces)| {
-                if let Some(piece) = pieces.last() {
-                    if piece.is_color(color) {
-                        return Some(*pos);
-                    }
-                }
-                None
-            })
-            .collect()
+    pub fn top_piece(&self, position: Position) -> Option<Piece> {
+        self.board.get(position).top_piece()
```

Comment:
This is confusing to me, top_piece calls another top_piece that isn't the same and takes no arguments? One of them should be renamed imo, this one could be top_piece_at.